Add sbt incremental compilation support

smarter opened the pull request:
Review by @DarkDimius

***
To test this with sbt, see
https://github.com/lampepfl/dotty/wiki/Using-Dotty-with-sbt

The following flags are added:
 - `-Yforce-sbt-phases`: Run the phases used by sbt for incremental compilation
   (ExtractDependencies and ExtractAPI) even if the compiler is ran outside of
   sbt, for debugging.
 - `-Ydump-sbt-inc`: For every compiled foo.scala, output the sbt API
   representation and dependencies in foo.inc, implies -Yforce-sbt-phases.

This commit introduces two new phases which do not transform trees:
- `ExtractDependencies` which extracts the dependency information of the current
  compilation unit and sends it to sbt via callbacks
- `ExtractAPI` which creates a representation of the API of the current compilation
  unit and sends it to sbt via callbacks

Briefly, when a file changes sbt will recompile it, if its API has
changed (determined by what `ExtractAPI` sent) then sbt will determine
which reverse-dependencies (determined by what `ExtractDependencies`
sent) of the API have to be recompiled depending on what changed.
See http://www.scala-sbt.org/0.13/docs/Understanding-Recompilation.html for
more information on how sbt incremental compilation works.

This phase was originally based on
https://github.com/adriaanm/scala/tree/sbt-api-consolidate/src/compiler/scala/tools/sbt
which attempts to integrate the sbt phases into scalac (and is itself based
on https://github.com/sbt/sbt/tree/0.13/compile/interface/src/main/scala/xsbt),
but it has been heavily refactored and adapted to Dotty. The main
functional differences are:
- ExtractDependencies runs right after Frontend (so that we don't lose
  dependency informations because of the simplifications done by PostTyper),
  but ExtractAPI runs right after PostTyper (so that SuperAccessors are
  part of the API).
- `ExtractAPI` only extract types as they are defined and never "as seen
  from" some some specific prefix, see its documentation for more details.
- `ExtractDependenciesTraverser` and `ExtractUsedNames` have been fused into
  one tree traversal in `ExtractDependenciesCollector`.

TODO: Try to run these phases in parallel with the rest of the compiler
pipeline since they're independent (except for the sbt callbacks in `GenBCode`) ?

smarter commented:
To run the scripted tests:
```shell 
git clone https://github.com/smarter/sbt.git -b inc-dotty
cd sbt
sbt "scripted source-dependencies/*"
```

DarkDimius commented on interfaces/src/main/java/dotty/tools/dotc/interfaces/CompilerCallback.java:
```diff
@@ -25,4 +27,12 @@
    */
   default void onSourceCompiled(SourceFile source) {};
+
+  default void sourceDependency(File dependsOn, File source, xsbti.DependencyContext context) {};
```
I dont see what a method does given its name. Should it be `addSourceDependency`?

smarter replied:
I'll remove these extra callbacks, they're not needed now that I added `Context#sbtCallback` instead.

DarkDimius commented on src/dotty/tools/dotc/core/SymDenotations.scala:
```diff
@@ -1060,6 +1060,8 @@ object SymDenotations {
       else defn.RootClass
     }
 
+    def isPublic(implicit ctx: Context): Boolean = !this.is(Protected | Private) && !privateWithin.exists
```
why do you need to check `privateWithin` if no `Protected | Private` is set?

DarkDimius replied:
why do you need to check `privateWithin` if no `Protected | Private` is set?

smarter replied:
Because we remove `Private` when a qualifier exists: https://github.com/lampepfl/dotty/blob/master/src/dotty/tools/dotc/parsing/Parsers.scala#L1422-L1427 `scalac` does the same thing. I guess this is because private qualified are not private in the bytecode, maybe we should do this in a miniphase later.

DarkDimius commented on src/dotty/tools/dotc/config/ScalaSettings.scala:
```diff
@@ -160,6 +160,8 @@ class ScalaSettings extends Settings.SettingGroup {
   val YcheckReentrant = BooleanSetting("-Ycheck-reentrant", "check that compiled program does not contain vars that can be accessed from a global root.")
   val YkeepComments = BooleanSetting("-Ykeep-comments", "Keep comments when scanning source files.")
+  val YforceSbtPhase = BooleanSetting("-Yforce-sbt-phase", "Run the sbt incremental compiler phase even if the compiler is ran outside of sbt, for debugging.")
+  val YdumpSbtApi = BooleanSetting("-Ydump-sbt-api", "For every compiled foo.scala, output the sbt API representation in foo.api, implies -Yforce-sbt-phase.")
```
`dependsOn(YforceSbtPhase)`?

smarter replied:
Looks like this would force the user to write `-Yforce-sbt-phase -Ydump-sbt-api`, I can do the change if you want but I don't think it's very useful

DarkDimius commented on src/dotty/tools/dotc/core/SymDenotations.scala:
```diff
@@ -1060,6 +1060,12 @@ object SymDenotations {
+    def isPublic(implicit ctx: Context): Boolean =
+      // private-qualified denotations do not have the Private flag set,
+      // see `ParserCommon#normalize`
+      !this.is(Protected | Private) && !privateWithin.exists
```
Why would you need to check for `privateWithin.exists` if neither `Protected`  nor `Private` flag is set?

smarter replied:
See https://github.com/lampepfl/dotty/pull/1244#discussion_r62421206

smarter replied:
This link behaves weirdly so copy-pasting:


Because we remove Private when a qualifier exists: https://github.com/lampepfl/dotty/blob/master/src/dotty/tools/dotc/parsing/Parsers.scala#L1422-L1427 scalac does the same thing. I guess this is because private qualified are not private in the bytecode, maybe we should do this in a miniphase later.

DarkDimius replied:
What you're saing that in your sbt-specific phase you have a different notion of public. 
For all other phases notion of `public` is simple: can a method\symbol\type be accessed from classes that are not subclasses of this class. Method that is `private[package]` is public from this definition. 

I'm against such change to either backend or SymDenotations.scala.

smarter replied:
A method that is `private[package]` is public in the bytecode but it's not public in the Scala language. `public` means accessible from everywhere, `private[package]` is only accessible in some cases (if you're in the same package).
Anyone who writes their own phase for dotty and wants to check if a symbol is public in the Scala language will want `isPublic` to return false for something which is `private[package]`, the fact that it is public in the bytecode is an implementation detail.

Anyway, I don't want to fight on this, I'll just put `def isPublic` in sbt and leave the rest alone.

smarter replied:
done.

DarkDimius commented on src/dotty/tools/backend/jvm/DottyBackendInterface.scala:
```diff
@@ -569,7 +569,7 @@ class DottyBackendInterface(outputDirectory: AbstractFile)(implicit ctx: Context
     def isAnonymousFunction: Boolean = toDenot(sym).isAnonymousFunction
     def isMethod: Boolean = sym is Flags.Method
-    def isPublic: Boolean =  sym.flags.is(Flags.EmptyFlags, Flags.Private | Flags.Protected)
+    def isPublic: Boolean =  toDenot(sym).isPublic
```
private [foo] methods should NOT be private on bytecode level.

smarter replied:
I agree, but note that `isPublic` in scalac is defined in the same way: https://github.com/scala/scala/blob/2.12.x/src/reflect/scala/reflect/internal/HasFlags.scala#L117
So I think the backend already takes this into account.

DarkDimius replied:
The common bakend does not.
I'm not sure why this is needed in scalac. I'm not in favor of such change in Dotty backend.
Nor do I think that we should add this incorrect implementation of `isPrivate` method: If method is defined as `private [package] def foo = ???` it has to be compiled as public. 
If some specific phase has different notion of `isPublic` - such implementation can go into this phase.

smarter commented:
I didn't realize this before but there's a single test that does not pass now: [`trait-super`](https://github.com/smarter/dotty-bridge/tree/incremental/src/sbt-test/source-dependencies/trait-super) because it relies on the presence or absence of superaccesors in the API to know what should be recompiled, but in dotty superaccessors are created by `PostTyper` but the sbt phase is run before `PostTyper` so that it can see constant types and trees of types (and so that the [`constants`](https://github.com/smarter/dotty-bridge/tree/incremental/src/sbt-test/source-dependencies/constants) test pass), I don't know if there's a way to fix this without doing some phase-travelling. Anyway, this is something that can be fixed later.

smarter commented:
OK I just fixed [`trait-super`](https://github.com/smarter/dotty-bridge/tree/incremental/src/sbt-test/source-dependencies/trait-super) by splitting the sbt phase into two phases, `ExtractDependencies` before `PostTyper` and `ExtractAPI` after `PostTyper` (they were already completely independent before, but ran in the same phase one after the other), this works out because:
- `ExtractDependencies` needs to be before `PostTyper` to avoid loosing dependency informations (see [`constants`](https://github.com/smarter/dotty-bridge/tree/incremental/src/sbt-test/source-dependencies/constants) for an example where this matters).
- `ExtractAPI` needs to be after `PostTyper` because SuperAccessors need to be part of the API (see again [`trait-super`](https://github.com/smarter/dotty-bridge/tree/incremental/src/sbt-test/source-dependencies/trait-super) for an example where this matter). Unlike `ExtractDependencies`, the simplication to trees done by `PostTyper` do not affect this phase because it only cares about definitions, and `PostTyper` does not change definitions.

I've also cleaned up the changes needed to `dotty-bridge`, they now sit in https://github.com/smarter/dotty-bridge/pull/1 and will be merged once this PR has been merged.

smarter commented:
/cc @gkossakowski and @adriaanm who might be interested to know that this exist (also thanks to both of you for the discussions we had on incremental compilation!)

DarkDimius commented:
Needs rebasing.

DarkDimius commented on src/dotty/tools/dotc/Compiler.scala:
```diff
@@ -41,56 +41,58 @@ class Compiler {
    */
   def phases: List[List[Phase]] =
     List(
-      List(new FrontEnd),           // Compiler frontend: scanner, parser, namer, typer
```
I'd prefer not to loose git history here.
Indentation is not that important.

Additionally, it will introduce a lot of merge conflicts with linker.

smarter replied:
OK, will revert formatting change (though note that `git blame -w` and `git diff -w` make this less important)

DarkDimius commented on src/dotty/tools/dotc/config/ScalaSettings.scala:
```diff
@@ -160,6 +160,8 @@ class ScalaSettings extends Settings.SettingGroup {
   val YcheckReentrant = BooleanSetting("-Ycheck-reentrant", "check that compiled program does not contain vars that can be accessed from a global root.")
   val YkeepComments = BooleanSetting("-Ykeep-comments", "Keep comments when scanning source files.")
+  val YforceSbtPhases = BooleanSetting("-Yforce-sbt-phases", "Run the phases used by sbt for incremental compilation (ExtractDependencies and ExtractAPI) even if the compiler is ran outside of sbt, for debugging.")
+  val YdumpSbtInc = BooleanSetting("-Ydump-sbt-inc", "For every compiled foo.scala, output the sbt API representation and dependencies in foo.inc, implies -Yforce-sbt-phases.")
```
it's not clear for me what does `Inc` stand for. The description does speak about it.

smarter replied:
inc stands for incremental, I'll make that explicit in the description

DarkDimius commented on test/dotc/tests.scala:
```diff
@@ -23,7 +23,7 @@ class tests extends CompilerTest {
 
   implicit val defaultOptions = noCheckOptions ++ List(
-      "-Yno-deep-subtypes", "-Yno-double-bindings",
+      "-Yno-deep-subtypes", "-Yno-double-bindings", "-Yforce-sbt-phases",
```
This is going to affect benchmarks and skew history.

smarter replied:
Yup, no choice.

DarkDimius commented on src/dotty/tools/dotc/sbt/ExtractAPI.scala:
```diff
@@ -0,0 +1,479 @@
+    val unit = ctx.compilationUnit
+    val dumpInc = ctx.settings.YdumpSbtInc.value
+    val forceRun = dumpInc || ctx.settings.YforceSbtPhases.value
+    if ((ctx.sbtCallback != null || forceRun) && !unit.isJava) {
```
What if someone changes java source? Don't you still need to rebuild?

smarter replied:
Java sources are handled differently by sbt, I don't actually know the details.

DarkDimius commented on src/dotty/tools/dotc/sbt/ThunkHolder.scala:
```diff
@@ -0,0 +1,59 @@
+   *  take lazy arguments when they could be strict, but this can be fixed in sbt,
+   *  see https://github.com/sbt/zinc/issues/114
+   */
+  def strictlzy[T <: AnyRef](t: T): api.Lazy[T] =
```
this method has unreadable name and is part of public interface.
Suggestion:
```
private def lazify...
```

smarter replied:
I'm confused by your suggestion, I can't have an overloaded method where the only difference is that the argument is `T` or `=> T`, they would conflict with each other.

`strictlzy` is only temporary anyway so its name is not super important, I could rename it to `strictLzy` which is easier to understand.

DarkDimius replied:
> I'm confused by your suggestion, I can't have an overloaded method where the only difference is that the argument is T or => T, they would conflict with each other.

Indeed. I was assuming that we can follow the code pattern in [Ensuring](https://github.com/scala/scala/blob/2.12.x/src/library/scala/Predef.scala#L315-L317). But without an argument this would indeed lead to ambiguous overload. 

>strictLzy

`strict2lzy` would be even easier to understand.

DarkDimius commented on src/dotty/tools/dotc/sbt/ThunkHolder.scala:
```diff
@@ -0,0 +1,59 @@
+import scala.collection.mutable.ListBuffer
+import xsbti.api
+
+/** Create and hold thunks */
```
It's not clear for me what are thunks.

DarkDimius replied:
This trait is very small and used only by `ExtractAPICollector`. It's not clear what it does. I propose to merge it into `ExtractAPICollector`

smarter replied:
A thunk is an unevaluated value that can be evaluated once, it's a well known term (especially used in Haskell, but not only). I can make the description longer to make this clearer.

This trait was intentionally split from `ExtractAPICollector` and I'd like to keep it this way:
- It has private mutable state that does not need to be modified by `ExtractAPICollector`
- The implementation is somewhat tricky, but it can be understood on its own without worrying about how it's used in `ExtractAPICollector`
- Similarly, someone can understand how `ExtractAPICollector` works without having to know the implementation details of `ThunkHolder`

DarkDimius commented on src/dotty/tools/dotc/sbt/ShowAPI.scala:
```diff
@@ -0,0 +1,156 @@
+// This file is copied straight from
```
as we always have sbt on classpath, isn't it simpler to simply use the one provided by sbt instead of maintaining a copy of it?

smarter replied:
When running dotty outside of sbt, we only have `"org.scala-sbt" % interface` (which only contains Java interfaces) on the classpath, `ShowAPI.scala` is in `"org.scala-sbt" % api` which will only be on the classpath when running dotty from sbt, but we want `-Ydump-sbt-inc` to work without running dotty from sbt to make debugging easier.

DarkDimius commented on src/dotty/tools/dotc/sbt/ExtractDependencies.scala:
```diff
@@ -0,0 +1,265 @@
+              traverse(tp.info)
+            traverse(tp.prefix)
+          }
+        case tp: ThisType =>
```
```
case tp: TypeProxy =>
  traverse(tp.underlying)
```
instead of all the cases?

smarter replied:
No, that would be incorrect. A `TypeBounds` is a `TypeProxy` whose `underlying` is the upper bound, but as the example in the documentation of this object shows, we need to record both the upper and lower bounds, `TypeTraverser#traverseChildren` will do that for us.

DarkDimius replied:
If the intended behaviour is special treatment of `TypeBounds` amoung all the  `TypeProxy` classes than let's make it explicit. At the moment it's implicitly hidden inside default case.
```scala

case tp: TypeBounds => 
  traverse(tp.hi); 
  traverse(tp.lo);
case tp: TypeProxy =>
  traverse(tp.underlying)
```

smarter replied:
I also had to redefine `SuperType` and `RefinedType`, I'm not convinced that copy-pasting behavior is better than relying on the default case, but I'll do the change anyway

smarter replied:
I'm reverting that change and going back to the previous version, it turns out that adding a case on all `TypeProxy` is dangerous because a `RefinedThis` is a `TypeProxy` and going to its underlying values results in infinite loops. I could add a special case for that but then we'll have even more special cases than the original version, even though the original version has some behavior which is implicit it's at least consistent with `NamedPartsAccumulator`

DarkDimius commented on src/dotty/tools/dotc/sbt/ExtractDependencies.scala:
```diff
@@ -0,0 +1,265 @@
+class ExtractDependencies extends Phase {
+  override def phaseName: String = "sbt-deps"
+
+  // This phase should be run directly after `Frontend`, if it is run after
```
this predicate can be checked by inspecting `ctx.phases` in `def run`

DarkDimius commented:
I've done a quick review, as I'm not into details of `xsbti.api`.

I'd be curious to see how much does introduction of this phase affect compilation time of dotty&stdlib.

Otherwise LGTM

smarter commented:
> I'd be curious to see how much does introduction of this phase affect compilation time of dotty&stdlib.

I spent a bunch of time optimizing it and currently it adds a few seconds to compilation, most of that time is just because of the extra tree traversal done by `ExtractDependenciesCollector` which I don't think we can avoid (I don't think `ExtractDependencies` can be fused with any other phase). It would be interesting to see if we could just run it in a separate thread.

liufengyun commented:
We have a  big improvement in performance, I guess it's due to incremental compilation:

https://d-d.me/tnc/dotty/web/

smarter commented:
That's very weird, the incremental compilation cannot possibily be used by these tests, they're not set up to be run through sbt and the dotty-bridge

Blaisorblade commented:
> ExtractAPI runs right after PostTyper (so that SuperAccessors are part of the API).

@smarter Can you confirm this prevents issues like https://github.com/sbt/sbt/issues/466, as I'm thinking?

To summarize: nowadays adding super calls to a `trait` requires recompiling implementing classes/objects, but SBT does not manage, because `superaccessors` is ignored by SBT (in https://github.com/sbt/sbt/issues/466#issuecomment-5918336 I wondered whether that could be fixed). It seems your design prevents such problems?
